### DISCUSSION TITLE
Connection Reset by Host Error

### DISCUSSION THREAD

**Discussion by christoph-mr** [sqlalchemy/sqlalchemy#10221] [Usage Questions]
Hi, I’m hoping somebody here can help me.  I have been working with a support engineer at MS and we have hit a wall.  I have been experiencing an intermittent (it occurs frequently enough to be an issue, but not all the time) connection reset issue:

`sqlalchemy.exc.OperationalError: (pyodbc.OperationalError) ('08S01', '[08S01] [Microsoft][ODBC Driver 18 for SQL Server]TCP Provider: Error code 0x2746 (10054) (SQLExecDirectW)')`

The error is usually triggered via simple select statements (e.g. 

```
<Table>.query.all()
<Table>.query.first()
<Table>.query.filter(…).all()
```   
Here are my connection strings:
```
SQLALCHEMY_DATABASE_URI = f'mssql+pymssql://{uid}:{pwd}@<Host>/<DB1>?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=No&TrustServerCertificate=Yes'
SQLALCHEMY_BINDS = {
'dbName':f'mssql+pyodbc://{uid}:{pwd}@<Host>/<DB2>?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=No&TrustServerCertificate=Yes'
	}
```
The nearest we can tell is that there is something at the network-level telling the SQL Server backend to close the connection due to how SQLAlchemy handles the underlying connections (e.g. thread pooling, connection pooling, etc.).  We see in the SNI Trace that SQL Server is closing the connection due to a _NetworkErrorFoundInInputStream_ error.  The source of the error is TDS.

**Setup**
I have a Flask web app (Python v.3.8.6) being hosted on a RHEL 8 server using gunicorn and nginx.  Here is the package info from the requirements file: 
```
Flask==2.3.2
Flask-SQLAlchemy==3.0.5
pyodbc==4.0.35
SQLAlchemy==2.0.19

```
The SQL Server instance (Microsoft SQL Server 2017 (RTM-CU31-GDR) (KB5021126) - 14.0.3460.9 (X64)) is being hosted on a Windows VM.  We also tried migrating the DBs to Azure SQL and, if anything, the issue happened more frequently.  

**What we've tried so far**
I’ve tried both ODBC Driver 17 for SQL Server and ODBC Driver 18 for SQL Server.  I’ve even recently tried using the FreeTDS driver (I receive a similar error from FreeTDS).  We’ve also looked at the SSL settings on Linux server:
```
MinProtocol = TLSv1.2
CipherString = DEFAULT:@SECLEVEL=1
```
We’ve played with various SQLAlchemy engine options (pool_pre_ping, isolation_level, different timeouts).  We’ve played with different driver settings (MarsConnection, Encryption, etc.)
We’ve run numerous bidtraces for the drivers and tcpdumps from the Linux server and haven’t found any issues.  
We’ve also looked at the SQL Server instance to verify that there aren’t any performance issues (we’ve run SNI Traces SSMS, profiler traces via SQL Server Profiler, and used MS’s SQLTrace). 
The most information that we’ve been able to gather is from the netsh trace on the SQL Server VM that reveals the _NetworkErrorFoundInInputStream_ error, however, we haven’t been able to identify what the underlying mechanism is that is triggering this.  I haven’t been able to recreate the error using straight pyodbc queries using pyodbc.connect.  Multiple pyodbc tests have succeeded without any issues.  I've also tried setting the pool class to NullPool, but no luck there either

**Comment by zzzeek** [comment]
hi -

how is gunicorn applying concurrency to the application, is it monkeypatching with gevent?  I dont think that would be compatible with pyodbc.

SQLAlchemy has no "thread pooling" and the connection pooling merely means that some connections are reused.

the stack trace here would be very important to see as this would indicate if the connection is being cut off when a. first connecting b. doing something like a pre-ping c. in the middle of a statement.   that's otherwise not shown here and the stack trace would show this.

**Reply by zzzeek** [reply]
but overall the most immediate red flag I see is gnuicorn, which does something exotic in order to get concurrency.  try deploying under mod_wsgi and see if that resolves.

**Reply by CaselIT** [reply]
Gunicorn has many different workers, the default ones are just process and thread based, so I don't think it's necessary the issue.

It has gevemt modes but that's entirely opt in

**Reply by zzzeek** [reply]
curious though that of the large list of options mentioned, nothing mentioned about gunicorn / nginx config at all, seems telling

**Reply by christoph-mr** [reply]
I haven't really specified many options for gunicorn:  
` --workers 4 --bind 0.0.0.0:8080 --log-level debug --timeout 600 --preload --reload `

There's a lot more going on in the nginx file though.  I could post some of that if it would be helpful.

**Reply by zzzeek** [reply]
what's `--reload` doing, is that doing some kind of dynamic module loading?

also how does the engine that's created using `create_engine()` get handed down to child forks?   is there a system in place to ensure that pyodbc connections aren't leaking across process boundaries?   (though using `NullPool` would be one way that should be more or less guaranteed)

**Reply by zzzeek** [reply]
does the error occur in developemnt, when you run nginix on a laptop and just do a single page / web request at a time?  or only under concurrency?  only on production?   only certain pages?  only after a few requests ?  does the error happen once, and then happens every time after that ?   or is it sporadic?   how often?  once an hour?  10 times a minute?   under what kind of concurrency?  etc.

**Reply by zzzeek** [reply]
does it happen only when a connection has been idle?   does SQL Server have any server side logging that it's dropping connections?  is there any kind of proxy between SQL Server and the ODBC client?  is there odbc pooling in use ?

**Reply by CaselIT** [reply]
> There's a lot more going on in the nginx file though. I could post some of that if it would be helpful.

I don't think nginx has any influence here, so it should not be needed.

>  --workers 4 --bind 0.0.0.0:8080 --log-level debug --timeout 600 --preload --reload

looking at this, try removing the preload (that does pre-fork and require a bit of care in sqlalchemy to support) and reload too just to be safe (but I think gunincorn reloads by killing a worker and re-starting it)

> I have been experiencing an intermittent (it occurs frequently enough to be an issue, but not all the time) connection reset issue:

expanding on what mike asked just above, how does the error presents itself? is it only after a while? or does it happen right away too, after the application starts

**Reply by christoph-mr** [reply]
It happens very often, but it's not predictable.  I would say closer to within minutes of using it, usually after a few page views.  It's on the order of minutes and not hours.  It happens in production.  Right now, there's only one user (me), because it's been pulled out of production.  No to proxy.  We've tried with odbc pooling enabled and disabled.

**Reply by CaselIT** [reply]
how are you testing that with pyodbc.connect it does not happen?
sqlalchemy with a null pool is basically calling pyodbc.connect each time you need a connection, so it does not make sense that it would not work

**Reply by christoph-mr** [reply]
From gunicorn on reload (https://docs.gunicorn.org/en/latest/settings.html?highlight=reload#reload):  

> Restart workers when code changes.

This setting is intended for development. It will cause workers to be restarted whenever application code changes.

The reloader is incompatible with application preloading. When using a paste configuration be sure that the server block does not import any application code or the reload will not work as designed.

The default behavior is to attempt inotify with a fallback to file system polling. Generally, inotify should be preferred if available because it consumes less system resources.

**Reply by christoph-mr** [reply]
I'll try remove the preload and reload parameters and see if they have any impact.

**Comment by christoph-mr** [comment]
> how are you testing that with pyodbc.connect it does not happen? sqlalchemy with a null pool is basically calling pyodbc.connect each time you need a connection, so it does not make sense that it would not work

I just wrote a little script to try querying the same DBs.

```
cnxn = pyodbc.connect('DRIVER={ODBC Driver 18 for SQL Server};SERVER='+server+';DATABASE='+db1+';ENCRYPT=No;TrustServerCertificate=Yes;Trusted_Connection=yes;')
cursor1 = cnxn.cursor()
cnxn2 = pyodbc.connect('DRIVER={ODBC Driver 18 for SQL Server};SERVER='+server+';DATABASE='+db2+';ENCRYPT=No;TrustServerCertificate=Yes;Trusted_Connection=yes;')
cursor2 = cnxn2.cursor()

if __name__ == "__main__":

    #print("starting for i = " + str(i))
    for j in range(300):
        q1 = cursor2.execute("Select * from Roles;")
        q2 = cursor2.execute("select top(1) * from Emergency_Department")
        q3 = cursor1.execute("Select * from Users where user='abcde';")
        #r1 = q1.fetchone()
        #r2 = q2.fetchone()
        #r3 = q3.fetchone()
        time.sleep(random.random())
        #print(str(i+1) + " completed")
    #print("finished for i = " + str(i))
    print("done")
```

**Reply by zzzeek** [reply]
sqlalchemy uses more features than that for pyodbc, including bound parameters, setinputsizes(), rollback() and commit().  also, try writing that same program above using SQLAlchemy , and run it the same way as you are running that pyodbc program.  it likely would not have any different results

**Reply by christoph-mr** [reply]
I'll give that a shot.

**Reply by CaselIT** [reply]
that's very strange that this works, since it's keeping the connections open, without closing / reopening them, so it should be the same as what sqlalchemy does.

if you try the same script using sqlalchemy does it work or you get the same error?
```py
engine = sa.create_engine('the-rul')
cnxn = engine.connect()
cnxn2 = engine.connect()

if __name__ == "__main__":

    #print("starting for i = " + str(i))
    for j in range(300):
        q1 = cnxn2.execute(sa.text("Select * from Roles;")).all()
        q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department")).all()
        q3 = cnxn.execute(sa.text("Select * from Users where user='abcde';")).all()
        #r1 = q1.fetchone()
        #r2 = q2.fetchone()
        #r3 = q3.fetchone()
        time.sleep(random.random())
        #print(str(i+1) + " completed")
    #print("finished for i = " + str(i))
    print("done")
```

**Reply by christoph-mr** [reply]
I'm double-checking the code now, but this is the traceback:

```
Traceback (most recent call last):
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
pyodbc.Error: ('HY000', '[HY000] [Microsoft][ODBC Driver 18 for SQL Server]Connection is busy with results for another command (0) (SQLExecDirectW)')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/svc-dev-tools/test2.py", line 31, in <module>
    q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department"))
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 483, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1635, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_single_context(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 2339, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/home/user/webapp/LinuxVE/lib/python3.8/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DBAPIError: (pyodbc.Error) ('HY000', '[HY000] [Microsoft][ODBC Driver 18 for SQL Server]Connection is busy with results for another command (0) (SQLExecDirectW)')
[SQL: select top(1) * from Emergency_Department]
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

**Reply by zzzeek** [reply]
OK so that's a completely different error.  that looks like you are sharing a pyodbc connection in some way, either concurrent use across two threads or maybe even using a connection for a new statement when it still has results pending from a previous statement.

does your application have appropriate `Session` handling in use, using a `Session` that is local to a single thread of execution / web request ?

**Reply by christoph-mr** [reply]
That was run outside of the web app.  That was the result of running the script with sqlalchemy.create_engine instead of pyodbc.connect.

**Reply by zzzeek** [reply]
what is the program ?  is this being run in gunicorn / nginix? what does a request handler look like?

**Reply by christoph-mr** [reply]
No, sorry, it was just a simple python script.

```
engine = sa.create_engine(f'mssql+pyodbc://{uid}:{pwd}@BMDCDWVDNA01/FLASK_USERS?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=No&TrustServerCertifcate=Yes')
engine2 = sa.create_engine(f'mssql+pyodbc://{uid}:{pwd}@BMDCDWVDNA01/CM_TRIAGE?driver=ODBC+Driver+18+for+SQL+Server&Encrypt=No&TrustServerCertifcate=Yes')

cnxn = engine.connect()
cnxn2 = engine2.connect()

if __name__ == "__main__":
    for j in range(300):
        q1 = cnxn2.execute(sa.text("Select * from Roles;"))
        q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department"))
        q3 = cnxn.execute(sa.text("Select * from Users where user='abcde';"))
        time.sleep(random.random())
    print("done")
```

**Reply by zzzeek** [reply]
wow so you run that script with "python script.py" and you get that stack trace?   so with your script above, please make sure you fetch all rows for each "q1" etc.   just calling .all() should be sufficient.

**Reply by christoph-mr** [reply]
Just updated it to call .all()   It didn't seem to have any effect.
```
if __name__ == "__main__":
    for j in range(300):
        q1 = cnxn2.execute(sa.text("Select * from Roles;"))
        q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department"))
        q3 = cnxn.execute(sa.text("Select * from Users where tso ='CH3RTSO';"))
        r1 = q1.all()
        r2 = q2.all()
        r3 = q3.all()
        time.sleep(random.random())
```

**Reply by zzzeek** [reply]
```py
if __name__ == "__main__":
    for j in range(300):
        q1 = cnxn2.execute(sa.text("Select * from Roles;"))
        q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department"))
        q3 = cnxn.execute(sa.text("Select * from Users where tso ='CH3RTSO';"))
        r1 = q1.all()
        r2 = q2.all()
        r3 = q3.all()
        time.sleep(random.random())
```

OK no, that's not correct, you are getting a second result set on `cnxn2` before you have consumed the previous one.  like this please:

```py
if __name__ == "__main__":
    for j in range(300):
        q1 = cnxn2.execute(sa.text("Select * from Roles;"))
        r1 = q1.all()
        q2 = cnxn2.execute(sa.text("select top(1) * from Emergency_Department"))
        r2 = q2.all()
        q3 = cnxn.execute(sa.text("Select * from Users where tso ='CH3RTSO';"))
        r3 = q3.all()
        time.sleep(random.random())
```

can you confirm that also fails?  on what iteration / statement does it fail, and is it deterministic?

**Reply by christoph-mr** [reply]
That worked.  Changing the order did the trick.

**Reply by christoph-mr** [reply]
And it ran without error.  Sorry, I should have mentioned that.

**Comment by christoph-mr** [comment]
I'm talking to the support engineer and he suggested that this may be network-related.  It's possible that the event though the appropriate ports/IPs are open it may be inspecting/altering the packets which is causing them to arrive out of order or malformed at the DB end and therefore is being closed by SQL Server.

**Reply by zzzeek** [reply]
it likely is network related but there are still app / driver level things that can affect this, such as sometimes I've seen if you do INSERTs with too many records, or very large BLOBs, etc., you suddenly get "network errors".  things like MTU etc. can be involved, all of that.   so trying to isolate the cause is important, like, when did this problem start and/or is it isolated to one environment out of others that don't have the problem, are the the biggest ones

**Reply by christoph-mr** [reply]
Sorry, I didn't read your full message.  This is occurring on a dev server.  It began occurring shortly after upgrading from RHEL7 to RHEL8.  There is a prod server that is still on RHEL7 that is running smoothly. 

```
Flask-SQLAlchemy==2.4.4
pyodbc==4.0.34
SQLAlchemy==1.3.19
```
